File: packages/control-contract/src/scope.ts
import type { Actor } from "./actor.js";
import type { Revision } from "./revision.js";
import type { TypedId } from "./ids.js";

/** Durable organization boundary for policy, identity, and budget. */
export type Organization = {
  readonly id: TypedId;
  readonly name: string;
  readonly revision: Revision;
  readonly created_at: string;
  readonly updated_at: string;
};

/** Goal ancestry is constrained to a single organization and is acyclic. */
export type Goal = {
  readonly id: TypedId;
  readonly organization_id: TypedId;
  readonly parent_goal_id: TypedId | null;
  readonly owner: Actor | null;
  readonly title: string;
  readonly description: string;
  readonly revision: Revision;
  readonly created_at: string;
  readonly updated_at: string;
};

export type ProjectStatus = "active" | "paused" | "archived";

/** An organization-scoped delivery context that can support several goals. */
export type Project = {
  readonly id: TypedId;
  readonly organization_id: TypedId;
  readonly goal_ids: readonly TypedId[];
  readonly name: string;
  readonly description: string;
  readonly status: ProjectStatus;
  readonly revision: Revision;
  readonly created_at: string;
  readonly updated_at: string;
};

/**
 * Durable project workspace identity. `local_path_hint` is mutable host
 * metadata, never its identity: moving a checkout must not make a new workspace.
 */
export type ProjectWorkspace = {
  readonly id: TypedId;
  readonly project_id: TypedId;
  readonly name: string;
  readonly repository_url: string | null;
  readonly local_path_hint: string | null;
  readonly revision: Revision;
  readonly created_at: string;
  readonly updated_at: string;
};

File: src-tauri/crates/altai-control-protocol/src/scope.rs
//! Canonical organization, goal, project, and workspace contracts.
//!
//! A workspace ID is durable project context. A local filesystem path is only
//! a host-specific locator and therefore cannot be used as its identity.

use crate::{Actor, GoalId, OrganizationId, ProjectId, Revision, WorkspaceId};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Organization {
    pub id: OrganizationId,
    pub name: String,
    pub revision: Revision,
    pub created_at: String,
    pub updated_at: String,
}

/// A goal belongs to exactly one organization. `parent_goal_id` is constrained
/// to that organization by the repository and must never form a cycle.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Goal {
    pub id: GoalId,
    pub organization_id: OrganizationId,
    pub parent_goal_id: Option<GoalId>,
    pub owner: Option<Actor>,
    pub title: String,
    pub description: String,
    pub revision: Revision,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProjectStatus {
    Active,
    Paused,
    Archived,
}

/// A project is organization-scoped and can support multiple goals.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Project {
    pub id: ProjectId,
    pub organization_id: OrganizationId,
    pub goal_ids: Vec<GoalId>,
    pub name: String,
    pub description: String,
    pub status: ProjectStatus,
    pub revision: Revision,
    pub created_at: String,
    pub updated_at: String,
}

/// Durable workspace/repository context for a project.
///
/// `local_path_hint` may change when a checkout moves and is deliberately not
/// part of workspace identity or de-duplication.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProjectWorkspace {
    pub id: WorkspaceId,
    pub project_id: ProjectId,
    pub name: String,
    pub repository_url: Option<String>,
    pub local_path_hint: Option<String>,
    pub revision: Revision,
    pub created_at: String,
    pub updated_at: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn workspace_path_is_not_its_identity() {
        let workspace = ProjectWorkspace {
            id: WorkspaceId::new("design-system"),
            project_id: ProjectId::new("website"),
            name: "Website checkout".to_string(),
            repository_url: Some("https://github.com/altaidevorg/altai-app".to_string()),
            local_path_hint: Some("/Users/a/src/altai-app".to_string()),
            revision: Revision::INITIAL,
            created_at: "2026-08-13T00:00:00Z".to_string(),
            updated_at: "2026-08-13T00:00:00Z".to_string(),
        };

        assert_eq!(workspace.id.value, "ws_design-system");
        assert_ne!(workspace.id.value, workspace.local_path_hint.unwrap());
    }
}
